### DISCUSSION TITLE
Improved workspace editing API

### DISCUSSION THREAD

**Discussion by isidorn** [microsoft/vscode-discussions#157] [Announcements]
#### 1. Editing notebooks with `WorkspaceEdit`

Workspace edits can now modify notebooks with `vscode.NotebookEdit`. The `vscode.NotebookEdit` class includes static factory functions for creating edits the insert, replace, and delete cells in a notebook:

```ts
const currentNotebook = vscode.window.activeNotebookEditor?.notebook;
if (currentNotebook) {
  const edit = new vscode.WorkspaceEdit();

  // Use .set to add one or more edits to the notebook
  edit.set(currentNotebook.uri, [
    // Create an edit that inserts one or more cells after the first cell in the notebook
    vscode.NotebookEdit.insertCells(/* index */ 1, [
      // ... new notebook cell data
    ]),

    // Additional notebook edits...
  ]);

  await vscode.workspace.applyEdit(edit);
}
```

#### 2. Performing snippet edits with `WorkspaceEdit`

The new `SnippetTextEdit`-type allows extensions to perform snippet edits with `WorkspaceEdit`. Snippet edits are similar to regular text edits, but they allow the user to tab through placeholders, insert variables, select from choice element etc. Learn more about snippets in [this guide](https://code.visualstudio.com/docs/editor/userdefinedsnippets#_snippet-syntax)

Note that snippet edits require a code or notebook editor and can only support snippet edits for a single editor, otherwise they behave like regular text edits.

We believe that combining snippet edits and workspace edits allow for new and more seamless code actions and refactorings, like the sample from the [GitHub Issues Notebook](https://marketplace.visualstudio.com/items?itemName=ms-vscode.vscode-github-issue-notebooks) extension below.

https://user-images.githubusercontent.com/1926584/194547192-7fe1169e-3ede-4d45-a3b8-a524c494651e.mp4

#### 3. Creating files with contents with `WorkspaceEdit`

You can now specify contents when creating files using `WorkspaceEdit`. Simply pass the contents with its options-bag, like so

```ts
const edit = new vscode.WorkspaceEdit();
edit.createFile(newFileUri, { contents: new Unit8Array([1, 2, 3]) });

await vscode.workspace.applyEdit(edit);
```

Try it out and let us know what you think! fyi @jrieken

**Comment by r3m0t** [comment]
Looks good!

The video for item 2 isn't loading

**Reply by isidorn** [reply]
Thanks, I just fixed the video.
